core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/component/CmVersionQueryService.java

```diff
@@ -0,0 +1,125 @@
+                .map(Package::getPkg)
+                .flatMap(List::stream)
+                .collect(Collectors.toMap(PackageName::getName, packageName -> Optional.ofNullable(packageName.getPattern())));
+        Map<String, List<PackageInfo>> fullPackageVersionsFromAllHosts = hostOrchestrator.getFullPackageVersionsFromAllHosts(gatewayConfig, packageMap);
+        LOGGER.debug("Reading CM package info, found packages: " + fullPackageVersionsFromAllHosts);
+        return fullPackageVersionsFromAllHosts;
+    }
+
+    PackageInfo checkCmPackageInfoConsistency(Map<String, List<PackageInfo>> cmPackageVersionsFromAllHosts) {
+        Multimap<String, PackageInfo> pkgVersionsMMap = HashMultimap.create();
+        cmPackageVersionsFromAllHosts.values()
+                .forEach(packageInfoList -> packageInfoList.forEach(
+                        packageInfo -> pkgVersionsMMap.put(packageInfo.getName(), packageInfo)));
+        validatePackageHasMultipleVersions(pkgVersionsMMap);
+        Set<PackageInfo> distinctPackageInfos = new HashSet<>(pkgVersionsMMap.values());
+        validatePackageInfoExists(distinctPackageInfos);
+        validateServerAndAgentHasSameVersion(distinctPackageInfos);
+        return distinctPackageInfos.stream().findFirst().get();
+    }
+
+    private void validateServerAndAgentHasSameVersion(Set<PackageInfo> distinctPackageInfos) {
+        Set<String> distinctVersions = distinctPackageInfos.stream()
+                .map(PackageInfo::getFullVersion)
+                .collect(Collectors.toSet());
+
+        if (distinctVersions.size() > 1) {
+            String error = "Error during sync! CM server and agent has different versions: "
+                    + distinctPackageInfos.stream()
+                    .map(PackageInfo::getPackageNameAndFullVersion)
+                    .collect(Collectors.joining(", "));
+            logAndThrowError(error);
+        }
+    }
+
+    private void validatePackageInfoExists(Set<PackageInfo> distinctPackageInfos) {
+        if (distinctPackageInfos.size() == 0) {
+            String error = "Error during sync! CM server and agent versions cannot be determined!";
+            logAndThrowError(error);
+        }
+    }
+
+    private void validatePackageHasMultipleVersions(Multimap<String, PackageInfo> pkgVersionsMMap) {
+        if (pkgVersionsMMap.keySet().size() < pkgVersionsMMap.size()) {
+            StringBuilder errorMessageBuilder = new StringBuilder(
+                    "Error during sync! The following package(s) has multiple versions present on the machines. ");
+            pkgVersionsMMap.asMap()
+                    .entrySet()
+                    .stream()
+                    .filter(packageInfos -> packageInfos.getValue().size() > 1)
+                    .forEach(packageInfos -> errorMessageBuilder
+                            .append("Package: ")
+                            .append(packageInfos.getValue()));
```

Comment:
you could use `collect( Collectors.joining("Package: "))` and then just concatenate the 2 string. a bit more clean and shorter in my opinion but sure does the same thing

Reply:
This is how I could refactor this part, do you consider it cleaner? It looks kindof similar to me..

            String baseErrorMessage = "Error during sync! The following package(s) has multiple versions present on the machines. Package: ";
            String packageErrorStr = pkgVersionsMMap.asMap()
                    .values()
                    .stream()
                    .filter(packageInfos -> packageInfos.size() > 1)
                    .map(Object::toString)
                    .collect(Collectors.joining("Package: "));
            String error = baseErrorMessage + packageErrorStr;

Reply:
I meant something like this:
```
        String packageErrorStr = pkgVersionsMMap.asMap()
                .values()
                .stream()
                .filter(packageInfos -> packageInfos.size() > 1)
                .map(Object::toString)
                .collect(Collectors.joining("Package: "));
        String error = "Error during sync! The following package(s) has multiple versions present on the machines. " + packageErrorStr;
```
Or the static part of the stirng could be a constant in the class. The variable in the method is not necessary, in that case:
```
        String packageErrorStr = pkgVersionsMMap.asMap()
                .values()
                .stream()
                .filter(packageInfos -> packageInfos.size() > 1)
                .map(Object::toString)
                .collect(Collectors.joining("Package: "));
       logAndThrowError(ERROR_MSG + packageErrorStr);
```
This way you would get rid off about 5 lines in the method compared to the original version